internal/storagenode/logstream/executor.go

```diff
@@ -647,22 +655,34 @@ func (lse *Executor) restoreLogStreamContext(rp storage.RecoveryPoints) *logStre
 		restoreMode = "recovered"
 		uncommittedLLSNBegin := cc.CommittedLLSNBegin + types.LLSN(cc.CommittedGLSNEnd-cc.CommittedGLSNBegin)
 		if uncommittedLLSNBegin-1 == last.LLSN {
-			lsc.storeReportCommitBase(cc.Version, cc.HighWatermark, uncommittedLLSNBegin, false)
+			uncommittedBegin := varlogpb.LogSequenceNumber{
+				LLSN: last.LLSN + 1,
+				GLSN: last.GLSN + 1,
+			}
+			lsc.storeReportCommitBase(cc.Version, cc.HighWatermark, uncommittedBegin, false)
 			lsc.uncommittedLLSNEnd.Store(uncommittedLLSNBegin)
-			lsc.setLocalLowWatermark(*first)
-			lsc.setLocalHighWatermark(*last)
+			lsc.setLocalLowWatermark(varlogpb.LogSequenceNumber{
+				LLSN: first.LLSN,
+				GLSN: first.GLSN,
+			})
 			return lsc
 		}
 	}
 
 	// something wrong
 	restoreMode = "invalid"
-	lsc.storeReportCommitBase(types.InvalidVersion, types.InvalidGLSN, types.InvalidLLSN, true)
+	lsc.storeReportCommitBase(types.InvalidVersion, types.InvalidGLSN, varlogpb.LogSequenceNumber{}, true)
 	if last != nil {
-		lsc.storeReportCommitBase(types.InvalidVersion, types.InvalidGLSN, last.LLSN+1, true)
+		uncommittedBegin := varlogpb.LogSequenceNumber{
+			LLSN: last.LLSN + 1,
+			GLSN: last.GLSN + 1,
```

Comment:
I understand it's for not saving hwm, but assigning GLSN to uncommittedBegin is awkward.

Reply:
I agree with you, and it looks awkward as well as confusing. As you said, the log stream can't know the following global log sequence number. 

In this implementation, by setting valid GLSN to the uncommittedBegin, the log stream can infer the local high watermark. On the other hand, when it is invalid, the log stream cannot guess the local high watermark because there is no log entry.